lib/Dialect/Triton/IR/Ops.cpp

```diff
@@ -365,18 +366,18 @@ mlir::LogicalResult mlir::triton::TransOp::inferReturnTypes(
     SmallVectorImpl<Type> &inferredReturnTypes) {
   // type is the same as the input
   auto argTy = operands[0].getType().cast<RankedTensorType>();
-  SmallVector<int64_t> retShape(argTy.getShape().begin(),
-                                argTy.getShape().end());
-  std::reverse(retShape.begin(), retShape.end());
+  auto order = properties.as<Properties *>()->order;
+  SmallVector<int64_t> retShape = applyPermutation(argTy.getShape(), order);
+
   auto retEltTy = argTy.getElementType();
   Attribute argEncoding = argTy.getEncoding();
   Attribute retEncoding;
   if (argEncoding) {
     Dialect &dialect = argEncoding.getDialect();
     auto inferLayoutInterface = dyn_cast<DialectInferLayoutInterface>(&dialect);
-    if (inferLayoutInterface->inferTransOpEncoding(argEncoding, retEncoding)
+    if (inferLayoutInterface
+            ->inferTransOpEncoding(argEncoding, order, retEncoding)
             .failed()) {
-      llvm::report_fatal_error("failed to infer layout for ReduceOp");
```

Comment:
Is there a reason why you removed this ?

Reply:
Fatal errors cause crashes rather than compile errors -- gdb was frustrated by this a few weeks ago.  Also this crash was causing the triton-lsp language server to crash.

It still returns a failure code on error?  Or are you concerned about losing the info printed to stderr?

Reply:
Yea just to stderr the reason of failure

Reply:
A problem with logging to stderr (with e.g. llvm::errs()) is that maybe this is an expected error for the caller, in which case we really *shouldn't* print anything.  An example again is the language server, where errors are totally expected.

Also if we just print out the one line ("failed to infer layout for ReduceOp"), it's not super helpful, like, *which* reduce op was failing?  During which phase of the compiler?  Etc.

I am not really sure how this is "supposed to work in mlir".  In absl, a failure status can have an error message associated with it, which the consumer can eventually choose to log or not.  But MLIR errors are just a boolean.

Reply:
Hmm I see, I haven't started using the language server, but yea that's annoying. Also, there's a lot of `llvm::report_fatal_error` in the code, are we gna remove all those ?

Yea I still don't know the best way to report errors in MLIR, but sometimes even a hint like "failed to infer layout for ReduceOp" could be useful. I don't think there's a way to tell exactly which op. The parser does print which line in the ttgir the error was encountered. 

Going back to the problem, if we actually remove the stderr, and we do hit an error, how can we tell what happened, does even gdb catch the mlir::failure() ? I don't think so, but I could be wrong.

Reply:
> Also, there's a lot of llvm::report_fatal_error in the code, are we gna remove all those ?

report_fatal_error is fine for things that "should never happen".  An IR parse failure is allowed to happen, though.

> Going back to the problem, if we actually remove the stderr, and we do hit an error, how can we tell what happened, does even gdb catch the mlir::failure() ? I don't think so, but I could be wrong.

The IR parser will print a very helpful error.  It says something like, this reduceop has the wrong encoding -- expected X but got Y.  "failed to infer layout for ReduceOp" doesn't really help in this case.

Reply:
> The IR parser will print a very helpful error. It says something like, this reduceop has the wrong encoding -- expected X but got Y. "failed to infer layout for ReduceOp" doesn't really help in this case.
oh, nevermind me then!